pkg/frontend/querymiddleware/cardinality.go

```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/grafana/dskit/cache"
+	"github.com/grafana/dskit/tenant"
+	otlog "github.com/opentracing/opentracing-go/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+const (
+	// cardinalityEstimateTTL is how long a key must see no write to expire and be removed from the cache.
+	cardinalityEstimateTTL = 7 * 24 * time.Hour
+
+	// cardinalityEstimateBucketSize is the size of buckets that queries are bucketed into.
+	cardinalityEstimateBucketSize = 24 * time.Hour
+
+	// cacheErrorToleranceFraction is how much the estimate must deviate
+	// from the actually observed cardinality to update the cache.
+	cacheErrorToleranceFraction = 0.1
+)
+
+// cardinalityEstimation is a Handler that caches estimates for a query's
+// cardinality based on similar queries seen previously.
+type cardinalityEstimation struct {
+	cache  cache.Cache
+	next   Handler
+	logger log.Logger
+
+	estimationError prometheus.Histogram
+}
+
+func newCardinalityEstimationMiddleware(cache cache.Cache, logger log.Logger, registerer prometheus.Registerer) Middleware {
+	estimationError := promauto.With(registerer).NewHistogram(prometheus.HistogramOpts{
+		Name:    "cortex_query_frontend_cardinality_estimation_difference",
+		Help:    "Difference between estimated and actual query cardinality",
+		Buckets: prometheus.ExponentialBuckets(10, 10, 6),
```

Comment:
A suggestion to have a more precise measurement. Rationale:
- Don't care a diff < 100 (it's negligible)
- With 10 buckets and a 2x factor we will have: 100, 200, 400, 800, 1.6k, 3.2k, 6.4k, 12.8k, 25.6k, 51.2k. Anything above 50k difference is a bad scenario, cause either we over or under estimated

```suggestion
		Buckets: prometheus.ExponentialBuckets(100, 2, 10),
```